libnavui-util/src/main/java/com/mapbox/navigation/ui/utils/internal/lifecycle/ViewLifecycleOwner.kt

```diff
@@ -0,0 +1,89 @@
+package com.mapbox.navigation.ui.utils.internal.lifecycle
+
+import android.view.View
+import androidx.annotation.UiThread
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleEventObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.LifecycleRegistry
+import androidx.lifecycle.lifecycleScope
+import androidx.lifecycle.repeatOnLifecycle
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import java.util.concurrent.Flow
+
+/**
+ * A [LifecycleOwner] that merges the [hostingLifecycleOwner]'s [Lifecycle] (of an Activity or Fragment)
+ * with the [View]'s drawing surface lifecycle.
+ *
+ * This lifecycle is useful for continuously running tasks that obtain the data and draw it in the [View],
+ * for example by collecting [Flow]s and drawing results in the nested views or on the map.
+ *
+ * In general, this lifecycle reflects the [hostingLifecycleOwner]'s [Lifecycle] whenever the `View` is attached
+ * but whenever the `View` is detached and we want to stop the coroutines tied to the lifecycle's scope,
+ * this lifecycle will transition to [Lifecycle.State.CREATED] (and invoke [Lifecycle.Event.ON_STOP])
+ * even if the hosting Activity or Fragment is still running.
+ *
+ * This lifecycle reaches [Lifecycle.State.DESTROYED] only when the [hostingLifecycleOwner] is destroyed.
+ *
+ * @see keepExecutingWhenStarted
+ */
+@UiThread
+class ViewLifecycleOwner(
+    private val hostingLifecycleOwner: LifecycleOwner
+) : LifecycleOwner {
+
+    private var isAttached = false
+
+    private val hostingLifecycleObserver = LifecycleEventObserver { _, event ->
+        val isAtLeastCreated = lifecycle.currentState.isAtLeast(Lifecycle.State.CREATED)
+        if (isAttached || (isAtLeastCreated && event == Lifecycle.Event.ON_DESTROY)) {
+            lifecycleRegistry.handleLifecycleEvent(event)
+        }
+    }
+    private val lifecycleRegistry = LifecycleRegistry(this)
+
+    init {
+        hostingLifecycleOwner.lifecycle.addObserver(hostingLifecycleObserver)
```

Comment:
if it the observer is added here, but never removed. won't that leak for activity/fragments that add and remove views?

or should this be in `onAttachedToWindow`, and removed in `onDetachedFromWindow`? 🤔

Reply:
This is only called in `View`'s `onCreate`. Removing and re-adding the same instance of a view won't add new lifecycle observers.

Reply:
Answering my own question here, no it won't leak because of the conditions in `hostingLifecycleObserver`.

It essentially, updates the lifecycle while attached and keeps it in the created state otherwise. When a destroy event occurs, this is also destroyed. 

So for view that are added and removed, they will stay in the "created" state as long as the hosting lifecycle is created. 

Awesome, that sounds good!

Reply:
Oh I needed to refresh github. Just noticed the comments above